schema/query/query.go

```diff
@@ -0,0 +1,72 @@
+package query
+
+import "github.com/rs/rest-layer/schema"
+
+type Query struct {
+	// Projection is the list of fields from the items of the result that should
+	// be included in the query response. A projected field can be aliased or
+	// given parameters to be passed to per field transformation filters. A
+	// projection is hierachical allow projection of deep structures.
+	//
+	// A DSL can be used to build the projection structure.
+	Projection Projection
+
+	// Predicate defines the criteria an item must meet in order to be
+	// considered for inclusion in the result set.
+	//
+	// A DLS can be used to build a predicate from a MongoDB like expressions.
+	Predicate Predicate
+
+	// Sort is a list of fields or sub-fields to use for sorting the result set.
+	Sort Sort
+
+	// Window defines result set windowing using an offset and a limit. When
+	// nil, the full result-set should be returned.
+	Window *Window
+}
+
+// New creates a query from a projection, predicate and sort queries using
+// their respective DSL notations. An optional window can be provided.
+//
+// Example:
+//
+//   New("foo{bar},baz:b", `{baz: "bar"}`, "foo.bar,-baz", Page(1, 10, 0))
+//
+// Select items with the foo field equal to bar, including only the foo.bar and
+// baz fields in the result set, with the baz field aliased to b. The result is
+// then sorted by the foo.bar field ascending and baz descending. The result is
+// windowed on page 1 with 10 items per page, skiping no result.
+func New(projection, predicate, sort string, window *Window) (*Query, error) {
```

Comment:
You're right, in all cases I need to incrementally build the query package. It might be better to have `New` creating an empty query and let the caller set the public fields.